### PR TITLE
MINOR: Update 4.0.0 documentation for Kafka Streams

### PR DESCRIPTION
**Pull request by aliehsaeedii**
This PR covers all the docs of Streams changes in release 4.0.0.

### CODE REVIEW

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -141,6 +141,22 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
 
     <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
 
+    <p>
+        In this release, eos-v1 (Exactly Once Semantics version 1) is no longer supported. Additionally, all deprecated methods, classes, APIs, and config parameters
+        up to (including) AK 3.5 release have been removed.
```
Comment (aliehsaeedii) [inline docs/streams/upgrade-guide.html]: Is listing all the removed APIs necessary?

**Comment by aliehsaeedii** [comment]
@mjsax 
all doc updates needed for 4.0.0?!

@bbejeck and @mjsax: I tried to write a summary of each KIP. Could you please check whether your KIPs are summarized good/accurate enough or some rewording is needed.

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (mjsax) [reply docs/streams/upgrade-guide.html]: Good question... It might be worth to highlight a few important ones like old Processor API, KStream#[flat]Transform[Values], KStream#through, KStream#branch,  Time/Session/SlidingWindows#of(), and KafkaStreams#setUncaughtExceptionHandler ?)

And it might be good to link to the Jira ticket for details: https://issues.apache.org/jira/browse/KAFKA-12822

For EOSv1, we should also call out that EOSv2 requires broker version 2.5+.

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+    <p>
+        Kafka Streams introduced the <code>RETRY</code> return option the <code>ProductionExceptionHandler</code> (<a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=311627309">KIP-1065</a>).
+        This feature addresses the issue of infinite retry loops caused by <code>TimeoutException</code> when an output topic might be missing.
+        Now the application is enabled to attempt the action again instead of either crashing or dropping records.
+        This change aims to enhance the robustness and reliability of Kafka Streams by providing a more flexible error-handling mechanism.
+    </p>
+
+    <p>
+
+    </p>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html:186]: This should go after the paragraph

File: docs/streams/upgrade-guide.html [hunk 3]
```diff
@@ -141,6 +141,22 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
 
     <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
 
+    <p>
+        In this release, eos-v1 (Exactly Once Semantics version 1) is no longer supported. Additionally, all deprecated methods, classes, APIs, and config parameters
+        up to (including) AK 3.5 release have been removed.
+    </p>
+
+    <p>
+        For better observability and performance, analysis are provided by additional metrics in Kafka clients and new methods added to the Admin,
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: Why do we have this paragraph? Seem all KIPs are listed individually below?

File: docs/streams/upgrade-guide.html [hunk 4]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
         <code>TransformerSupplier</code>, <code>ValueTransformer</code>, and <code>ValueTransformerSupplier</code>.
     </p>
 
+    <p>
+        Kafka Streams introduced the <code>RETRY</code> return option the <code>ProductionExceptionHandler</code> (<a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=311627309">KIP-1065</a>).
+        This feature addresses the issue of infinite retry loops caused by <code>TimeoutException</code> when an output topic might be missing.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        This feature addresses the issue of infinite retry loops caused by <code>TimeoutException</code>, for example when an output topic does not exist.
```

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+        Now the application is enabled to attempt the action again instead of either crashing or dropping records.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: > Previously, the `ProductionExceptionHandler` was not invoked on a (retriable) `TimeoutException`. With Kafka Streams 4.0, the handler is called, and the default handler would return `RETRY` to not change existing behavior. However, a custom handler can now decide to break the infinite retry loop by returning either `CONTINUE` or `FAIL`.

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+        This change aims to enhance the robustness and reliability of Kafka Streams by providing a more flexible error-handling mechanism.
+    </p>
+
+    <p>
+
+    </p>
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1076%3A++Metrics+for+client+applications+KIP-714+extension">KIP-1076</a>
+        proposes adding methods to the Admin, Consumer, and Producer interfaces to allow applications embedding Kafka clients to register and push application-specific metrics,
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
       adds new methods <code>registerMetricForSubscription(KafkaMetric)</code> and <code>unregisterMetricFromSubscription(KafkaMetric)</code> to the Admin, Consumer, and Producer interfaces to allow applications embedding Kafka clients to register application-specific metrics for pushing them to the broker,
```

File: docs/streams/upgrade-guide.html [hunk 7]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+        supplementing the client metrics introduced in KIP-714.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: > the client metrics introduced in

KIP-714 did not introduce new metrics IIRC (and even if, it would not be the point, right?) -- the KIP is about collection client metrics broker side.

File: docs/streams/upgrade-guide.html [hunk 8]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+
+    <p>
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1077%3A+Deprecate+%60ForeachProcessor%60+and+move+to+internal+package">KIP-1077</a>
+        deprecates the <code>ForeachProcessor</code> class and move it to an internal package.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        deprecates the <code>ForeachProcessor</code> class.
```

Comment (mjsax) [reply docs/streams/upgrade-guide.html]: No important, and an impl detail that we keep it in an internal package.

File: docs/streams/upgrade-guide.html [hunk 9]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+        This change is aimed at improving the organization and clarity of the Kafka Streams API by ensuring that internal classes are not exposed in public packages.
+    </p>
+
+    <p>
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1078%3A+Remove+Leaking+Getter+Methods+in+Joined+Helper+Class">KIP-1078</a> deprecates the leaking getter methods in the <code>Joined</code> helper class.
+        These methods should be deprecated for future removal as they are already available via <code>JoinedInternal</code> class.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        These methods are deprecated without a replacement for future removal, as they don't add any value to Kafka Streams users.
```

File: docs/streams/upgrade-guide.html [hunk 10]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+
+    <p>
+        To ensures better encapsulation and organization of configuration documentation within Kafka Streams,
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1085%3A+Fix+leaking+*_DOC+variables+in+StreamsConfig">KIP-1085</a>
+        deprecate certain public doc description variables that are only used within the <code>StreamsConfig</code> or <code>TopologyConfig</code> classes.
+        Additionally, the unused variable <code>DUMMY_THREAD_INDEX</code> will also be deprecated.
+    </p>
+
+    <p>
+        Due to the recent deprecation of the <code>#through</code> method in Kafka Streams, the <code>intermediateTopicsOption</code> of <code>StreamsResetter</code> tool in Apache Kafka is
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        Due to the removal of the already deprecated <code>#through</code> method in Kafka Streams, the <code>intermediateTopicsOption</code> of <code>StreamsResetter</code> tool in Apache Kafka is
```

File: docs/streams/upgrade-guide.html [hunk 11]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+        not needed any more and therefore is deprecated (<a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1087%3A+Removing+intermediateTopicsOption+from+StreamsResetter">KIP-1087</a>).
+    </p>
+
+    <p>
+        In this release, explicit status information about the application's health is provided by new numeric metrics for both client-level and thread-level states
+        (<a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1091%3A+Improved+Kafka+Streams+operator+metrics">KIP-1091</a>).
+        These metrics will be available at the <code>INFO</code> recording level, and a thread-level metric with a String value will be available for users leveraging Java Management Extensions (<code>JMX</code>).
+    </p>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html:215]: Might be good to mention the motivation? The existing metrics cannot be collected via KIP-714, so we introduce numeric counterparts to allow proper broker side metric collection for KS apps.

File: docs/streams/upgrade-guide.html [hunk 12]
```diff
@@ -161,6 +177,65 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
...
+
+    <p>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html:217]: Not a KS KIP. I would remove this paragraph.

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        up to (including) Kafka Streams 3.6 release have been removed.
```

Comment (mjsax) [reply docs/streams/upgrade-guide.html]: We keep 3 release of history, so 3.9/3.8/3.7, right?

I guess, we did not deprecate anything in 3.6, so nothing needed to be removed, but I would still say 3.6 here?

**Review by mjsax — COMMENTED** [review]
Thanks. Made a pass.

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (aliehsaeedii) [reply docs/streams/upgrade-guide.html]: I assumed `Selected KIPs (maybe apply a grouping) should be added to the top level AK upgrade guide`

Comment (aliehsaeedii) [reply docs/streams/upgrade-guide.html]: I removed it anyway.

**Comment by aliehsaeedii** [comment]
@mjsax Final question: I assume KIP-1106 should not be listed here. Right?

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (mjsax) [reply docs/streams/upgrade-guide.html]: > to the top level AK upgrade guide

Yes, but this is not the "top level" page -- The top level page is `docs/upgrade.html` -- it contains a section for each Kafka release, including bullet points for selected KIPs. The only thing worth high-lighting there, might be KIP-1076?

In this KS specific sub-page, we list all KS KIPs.

**Comment by mjsax** [comment]
> Final question: I assume KIP-1106 should not be listed here. Right?

I think we should include it, because is does contain API changes for KS. The `enum Topology.AutoOffsetReset` was deprecated and replaced with a new `class AutoOffsetReset`.

File: docs/streams/upgrade-guide.html [hunk 13]
```diff
@@ -141,6 +141,20 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
 
     <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
 
+    <p>
+        In this release, eos-v1 (Exactly Once Semantics version 1) is no longer supported. To use eos-v2, brokers must be running version 2.5 or later.
+        Additionally, all deprecated methods, classes, APIs, and config parameters up to and including AK 3.5 release have been removed.
+        A few important ones are listed below. The full list can be found <a href="https://issues.apache.org/jira/browse/KAFKA-12822">here</a>.
+        <ul>
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12829">Old processor APIs</a></li>
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12823">kstream.KStream#through in both Java and Scala</a></li>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12823">KStream#through() in both Java and Scala</a></li>
```

File: docs/streams/upgrade-guide.html [hunk 14]
```diff
@@ -141,6 +141,20 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
...
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-16339">"transformer" methods and classes in both Java and Scala</a></li>
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12824">kstream.KStream#branch in both Java and Scala</a></li>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html:152]:
```suggestion
            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12824">KStream#branch() in both Java and Scala</a></li>
```

File: docs/streams/upgrade-guide.html [hunk 15]
```diff
@@ -141,6 +141,20 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
...
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-16332">builder methods for Time/Session/Join/SlidingWindows</a></li>
+            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12827">KafkaStreams#setUncaughtExceptionHandler</a></li>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
            <li><a href="https://issues.apache.org/jira/browse/KAFKA-12827">KafkaStreams#setUncaughtExceptionHandler()</a></li>
```

**Review by mjsax — COMMENTED** [review]
Overall LGTM. 3 nits.

Also let's add a bullet point for KIP-1076 to the top-level upgrade guide (I would rephrase the text though, only saying, "With KIP-1076, Kafka Streams metrics can be collected broker side via the KIP-714 broker-plugin." -- that's the main change we ship.

The fact that we added consumer/producer/admin method to register user metrics for collections, is not too important IMHO (if you want to add it in addition, also ok with me), but rather a impl details (at least from KS POV).

**Comment by aliehsaeedii** [comment]
> Also let's add a bullet point for KIP-1076 to the top-level upgrade guide (I would rephrase the text though, only saying, "With KIP-1076, Kafka Streams metrics can be collected broker side via the KIP-714 broker-plugin." -- that's the main change we ship.

What do you mean by `add a bullet point for KIP-1076 to the top-level upgrade guide`? The KIPs are listed based on the kip-number. Do you mean I add an extra line/paragraph for that (caz it's important) or just change the order and list it as the first one?
BTW, I rephrased the text as you suggested.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `4.0` branch.
